Rewrite, remove inaccurate Range truncation info

jstuder-gh opened the pull request:
Previously, this section stated that:

* All ranges (including non-lazy ones) will truncate the output/ omit
  undefined elements.

Rewrote the section to remove that inaccuracy and address the following:

* If the element at index [x] does not exist then the result will be an
  undefined element (when subscripting Positionals).
     * This behavior can be modified using the :v adverb.
* Using lazy Iterables as subscripts yields a different behavior. Only
  returns defined elements (most of the time; see bullets below).
    * Behind the scenes the elements of the lazy subscript are being
      reified up until the point that the element in the Positional
      is undefined.
    * Continued use of lazy Lists will begin to return undefined
      elements as new elements in the lazy List are reified.
* Include justification for these behaviors. Sometimes returning
  undefined elements is desirable.
    * DESIRABLE:     assignment. eg @a[^10] = 1..10;
    * NOT DESIRABLE: Out-of-memory due to uncontrolled indexing.

This removes some information about lack of protection against runaway
slices and reification of subscript elements in instances where the
undefined values are not being returned from the collection (eg,
instances where the same index is being used over and over again). This
seemed a little out of place in this section and probably fits better
elsewhere. Perhaps it would be appropriate in the Traps section?

SEE ALSO:
    [Issue #1679](https://github.com/perl6/doc/issues/1679)
    https://irclog.perlgeek.de/perl6/2017-11-18#i_15466865

jstuder-gh commented:
Looking for review from more senior members as this is a substantial rewrite of this section. Thanks.

AlexDaniel commented:
```perl6
my @letters = <a b c d e f>; @letters[4]:delete; say @letters[3..*]
```
```
(d (Any) f)
```
```perl6
my @letters = <a b c d e f>; @letters[4]:delete; say @letters[lazy 3..*]
```
```
(d)
```
I always thought that this was a bug… I guess this PR documents the current behavior, right?

zoffixznet commented on doc/Language/subscripts.pod6:
```diff
@@ -253,68 +251,74 @@ dimensions>) the subscript is preserved across the slice operation
-will stop if it generates a value that points to a nonexistent index.
+This behavior, while at first glance may seem unintuitive, is desirable in
+instances when you want to assign a value at an index in which a defined
+value does not currently exist.
```
*[request for change]*

> in which a defined value

"defined" should be dropped here, since [undefined](https://docs.perl6.org/routine/defined) are perfectly acceptable elements as far as existence goes.

zoffixznet commented on doc/Language/subscripts.pod6:
```diff
@@ -253,68 +251,74 @@ dimensions>) the subscript is preserved across the slice operation
-the indices.  As such, the following will most likely hang until all
-memory has been consumed:
+If you want the resulting slice to only include defined elements, you can
+silently skip the non-defined elements using the L<#:v> adverb.
```
*[request for change]*

Same as above, change "defined" to "existent" ("extant"?), since it doesn't care about undefinedness:

```
$ perl6 -e '(Any, Int)[1..*]:v.say'
((Int))
```

zoffixznet commented on doc/Language/subscripts.pod6:
```diff
@@ -253,68 +251,74 @@ dimensions>) the subscript is preserved across the slice operation
+result.
+
+Behind the scenes, the implementation is first eagerly accessing the collection
+via subscripts that are already L<reified|/language/glossary#index-entry-Reify>.
```
*[feedback]*

IMO this paragraph goes into too much detail:

* Most users shouldn't care about these details and shouldn't need to read all this extra info, at least in this document
* This describes how *Rakudo* implemented this feature, while the docs site tries to be implementation-agnostic and only describe what [the spec](https://github.com/perl6/roast/) specifies the behaviour to be. I don't think the spec describes such minute details as when to reify things in this particular scenario.

zoffixznet commented on doc/Language/subscripts.pod6:
```diff
@@ -253,68 +251,74 @@ dimensions>) the subscript is preserved across the slice operation
-So, to safely use lazy indices, they should be one-shot things which
-are guaranteed to overrun the array.  The following alternate formulation
-will produce a fully lazy result (but will not truncate):
+The behavior when indexing a collection via lazy subscripts is different than
```
*[request for change]*

Would probably be helpful to link `"lazy"` to https://docs.perl6.org/routine/is-lazy or something similar that explains what the term means.

zoffixznet commented on doc/Language/subscripts.pod6:
```diff
@@ -253,68 +251,74 @@ dimensions>) the subscript is preserved across the slice operation
+subscript reification is halted and the resulting slice is returned.
+
+If you are subscripting via a lazy C<List>, be aware that each subsequent use
+reifies and caches an additional element in the lazy C<List>, introducing
```
*[feedback]*

This problem too feels implementation-specific IMO and probably shouldn't be listed here. I opened a ticket for this problem in Rakudo: https://github.com/rakudo/rakudo/issues/1320 

It likely won't get fixed and in such a case a better place for this might be in [Reification](https://docs.perl6.org/language/glossary#index-entry-Reify) docs or in [Traps](https://docs.perl6.org/language/traps).

jstuder-gh commented:
@zoffixznet,
I've made the changes to the section that you requested in your review. I added the "drift" explanation to Traps and a link to a "Lazy List" section to explain the context of laziness a bit.

I made a few changes to the "Lazy List" section I linked to which is outside the scope of this PR, but I can remove that if you prefer the old language better. I added an index to the section as an explanation of the general concept of laziness. Let me know if you want that removed.

I had been holding off on updating this branch as I wasn't sure how to document lazy indexing and how it reacts to holes, but I suppose we can update the section in the future to address holes whichever way we decide to proceed on that issue. 

I rebased the branch so that its base is a more recent commit. I believe this is ready to be merged in now, assuming the recent revisions have addressed your concerns. Let me know what you think.

zoffixznet commented:
👍 Thanks!
